staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1201,41 +1296,78 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
 func (c *cacheWatcher) stopLocked() {
 	if !c.stopped {
 		c.stopped = true
-		close(c.done)
 		close(c.input)
 	}
+
+	// stop without draining the input channel was requested
+	// make sure the done channel is not already closed otherwise it will panic
+	if !c.drainInputBuffer && !c.isDoneChannelClosed() {
+		close(c.done)
+	}
 }
 
 func (c *cacheWatcher) nonblockingAdd(event *watchCacheEvent) bool {
+	// if the bookmarkAfterResourceVersion hasn't been seen
+	// we will try to deliver a bookmark event every second.
+	// the following check will discard a bookmark event
+	// if it is < than the bookmarkAfterResourceVersion
+	// so that we don't pollute the input channel
+	if event.Type == watch.Bookmark && event.ResourceVersion < c.bookmarkAfterResourceVersion {
+		return false
+	}
 	select {
 	case c.input <- event:
+		c.markBookmarkAfterResourceVersionAsReceived(event)
 		return true
 	default:
 		return false
 	}
 }
 
 // Nil timer means that add will not block (if it can't send event immediately, it will break the watcher)
+//
+// Note that bookmark events are never added via the add method only via the nonblockingAdd.
+// Changing this behaviour will require moving the markBookmarkAfterResourceVersionAsReceived method
 func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 	// Try to send the event immediately, without blocking.
 	if c.nonblockingAdd(event) {
 		return true
 	}
 
 	closeFunc := func() {
-		// This means that we couldn't send event to that watcher.
-		// Since we don't want to block on it infinitely,
-		// we simply terminate it.
-		klog.V(1).Infof("Forcing %v watcher close due to unresponsiveness: %v. len(c.input) = %v, len(c.result) = %v", c.objectType.String(), c.identifier, len(c.input), len(c.result))
 		metrics.TerminatedWatchersCounter.WithLabelValues(c.objectType.String()).Inc()
-		c.forget()
+		// This means that we couldn't send event to that watcher.
+		// Since we don't want to block on it infinitely, we simply terminate it.
+
+		// we are graceful = false, when:
+		//
+		// (a) The bookmarkAfterResourceVersionReceived hasn't been received,
+		//     we can safely terminate the watcher. Because the client is waiting
+		//     for this specific bookmark, and we even haven't received one.
+		// (b) We have seen the bookmarkAfterResourceVersion, and it was sent already to the client.
+		//     We can simply terminate the watcher.
+
+		// we are graceful = true, when:
+		//
+		// (a) We have seen a bookmark, but it hasn't been sent to the client yet.
+		//     That means we should drain the input buffer which contains
+		//     the bookmarkAfterResourceVersion we want. We do that to make progress
+		//     as clients can re-establish a new watch with the given RV and receive
+		//     further notifications.
+		graceful := func() bool {
+			c.stateMutex.Lock()
+			defer c.stateMutex.Unlock()
+			return c.wasBookmarkAfterResourceVersionReceivedLocked() && !c.wasBookmarkAfterResourceVersionSent()
```

Comment:
> we are graceful when we have received an RV but i hasn't been sent, so we cannot simply check graceful := (c.state == cacheWatcherBookmarkReceived)

I don't understand. The semantic of cacheWatcherBookmarkReceived is exactly that:
- we already received a bookmark (with that RV)
- but we didn't send that yet

Reply:
`graceful` indicates whether we should drain the input chan or not. 

we are draining the input chan only when
1.  we have received a bookmark with an RV >= the expected one
2. the bookmark hasn't been sent to the client

The semantic of `cacheWatcherBookmarkReceived` is `we already received a bookmark (with that RV)`, given the above we also need to check `2`

Reply:
> The semantic of cacheWatcherBookmarkReceived is we already received a bookmark (with that RV), given the above we also need to check 2

No - as soon as we send the bookmark we call `markBookmarkAfterResourceVersionSentLocked` which is changing the state. So we actually are in this state exactly during that period.
This is the whole reason why I wanted state-machine - it's just very explicit.